File: 2021_summer_algorithm_study_intermediate/using_python/10891_memout.py
"""
input :
4 4
1 2
2 3
3 1
3 4

output :
Cactus
"""
import sys
sys.stdin = open("input.txt", "r")
input = sys.stdin.readline
sys.setrecursionlimit(10 ** 5)


def dfs(p, mp):
    global NODE_NUM; global BCC_NUM;
    NODE_NUM += 1
    dfn[p] = low[p] = NODE_NUM

    for chd in graph[p]:
        if chd == mp: continue

        if dfn[chd] == 0:  # 처음 진입
            ST.append([p, chd])
            child[p] += 1
            dfs(chd, p)
            low[p] = min(low[p], low[chd])

            if dfn[p] <= low[chd]:
                if mp != -1:
                    AC[p] = True

                BCC_NUM += 1
                while True:
                    a, b = ST.pop()
                    BCC[BCC_NUM].append([a, b])
                    if [a, b] == [p, chd]: break

        elif dfn[chd] < dfn[p]:
            ST.append([p, chd])
            low[p] = min(low[p], dfn[chd])

    if mp == -1 and child[p] >= 2:
        AC[p] = True


def get_node_num(edges):
    chk_set = set()
    for edge in edges:
        chk_set.add(edge[0])
        chk_set.add(edge[1])

    return len(chk_set)


if __name__ == "__main__":
    V, E = map(int, input().rstrip().split())
    graph = dict()
    for i in range(1, V + 1):
        graph[i] = list()

    for _ in range(E):
        a, b = map(int, input().rstrip().split())
        graph[a].append(b)
        graph[b].append(a)

    BCC = [[] for _ in range(V + 1)]
    dfn = [0 for _ in range(V + 1)]
    low = [0 for _ in range(V + 1)]
    child = [0 for _ in range(V + 1)]
    AC = [False for _ in range(V + 1)]
    ST = list()
    NODE_NUM = 0
    BCC_NUM = 0
    for i in range(1, V + 1):
        if dfn[i]: continue

        dfs(i, -1)

    # < 선인장일 조건 >
    # 크기가 3 이상인 BCC이면, 간선개수 = 정점개수
    # 각 정점별로 크기가 3 이상인 BCC는 하나 이하로만 있어야함
    nd2bccnums = dict()
    for i in range(1, V + 1):
        nd2bccnums[i] = set()

    for b_num in range(1, BCC_NUM + 1):
        for edge in BCC[b_num]:
            nd2bccnums[edge[0]].add(b_num)
            nd2bccnums[edge[1]].add(b_num)

    for nd in nd2bccnums:
        cnt = 0
        for b_num in nd2bccnums[nd]:
            if len(BCC[b_num]) >= 3:
                cnt += 1
                if cnt > 1:  # 하나의 정점에 '크기3이상의BCC'가 여러개 걸려있으면 안된다. (많아야 1개)
                    print("Not cactus")
                    exit()
                if len(BCC[b_num]) != get_node_num(BCC[b_num]):  # '크기3이상의BCC'가 달려 있으려면, 필히 간선 개수 == 정점 개수 여야함
                    print("Not cactus")
                    exit()
    print("Cactus")

File: 2021_summer_algorithm_study_intermediate/using_cpp/10891_memout.cpp
/*
input :
4 4
1 2
2 3
3 1
3 4

output :
Cactus
*/

#include <iostream>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#define SZ (100005)
using namespace std;

typedef struct edge{
    int a, b;
}edge;


// 전역 변수 선언
int V, E;
vector<int> graph[SZ];
vector<edge> BCC[SZ];
int dfn[SZ], low[SZ];
vector<edge> ST;
int NODE_NUM, BCC_NUM;

void dfs(int p, int mp = -1){
    dfn[p] = low[p] = ++NODE_NUM;
    for (int chd: graph[p]){
        if (chd == mp) continue;  // 부모로 가는 tree edge는 제외. 아래로만 갈 수 있도록.

        if (dfn[chd] == 0){ // 첫 방문
            ST.push_back({p, chd});
            dfs(chd, p);
            low[p] = min(low[p], low[chd]);

            // BCC 생성
            if (dfn[p] <= low[chd]){
                BCC_NUM++;
                while (true){
                    edge ab = ST.back(); ST.pop_back();
                    BCC[BCC_NUM].push_back(ab);
                    if (ab.a == p and ab.b == chd) break;
                }
            }
        }
        else if (dfn[p] > dfn[chd]){ // 조상 touch
            low[p] = min(low[p], dfn[chd]);
            ST.push_back({p, chd});
        }
    }
}

int get_node_num(vector<edge> edges){
    set<int> chk_set;
    for (edge &ab: edges){
        chk_set.insert(ab.a);
        chk_set.insert(ab.b);
    }
    return chk_set.size();
}


int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    freopen("input.txt", "r", stdin);

    cin >> V >> E;
    for (int i = 1; i <= E; i++){
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    for (int i = 1; i <= V; i++){
        if (dfn[i]) continue;

        dfs(i, -1);
    }

    // print ans
    map<int, set<int>> nd2bccnums;
    for (int i = 1; i <= V; i++){
        nd2bccnums[i] = {};
    }
    for (int b_num = 1; b_num <= BCC_NUM; b_num++){
        for (edge &ab: BCC[b_num]){
            nd2bccnums[ab.a].insert(b_num);
            nd2bccnums[ab.b].insert(b_num);
        }
    }
    for (int nd = 1; nd < V; nd++){
        int cnt = 0;

        for (int b_num: nd2bccnums[nd]){
            if (BCC[b_num].size() >= 3){
                cnt += 1;
                if (cnt > 1){
                    cout << "Not cactus" << '\n';
                    return 0;
                }
                if (BCC[b_num].size() != get_node_num(BCC[b_num])){
                    cout << "Not cactus" << '\n';
                    return 0;
                }
            }
        }
    }
    cout << "Cactus" << '\n';

    return 0;
}